### PR TITLE
Write down precise constants in the asymptotic complexity

### PR DESCRIPTION
**Pull request by girving**
1. Alice takes 5000 k^2 t log t + smaller terms
2. Bob takes 20000/9 k^2 t log t + smaller terms
3. Vera takes 106000 * k^2 + 1

### CODE REVIEW

**Comment by girving** [comment]
@eric-wieser Next one: A bit of refactoring, and then writing down the constants.

**Comment by girving** [comment]
(For more context: this is very much related to the original paper, and will be the last PR for a while based on recent discussions.)

File: Debate/Cost.lean [hunk 1]
```diff
@@ -109,6 +111,44 @@ theorem bob_debate_cost (o : Oracle) (alice : Alice) (vera : Vera) (t : ℕ):
   · refine exp_le_of_forall_le (fun y m ↦ ?_)
     induction y; repeat simp only [Comp.cost_pure, le_refl]
 
+/-- `Nat.ceil` adds at most one -/
+lemma Nat.ceil_le_add_one {x : ℝ} (x0 : 0 ≤ x) : ⌈x⌉₊ ≤ x + 1 := by
+  rw [natCast_ceil_eq_intCast_ceil x0]
+  exact (Int.ceil_lt_add_one x).le
```
Comment (eric-wieser) [inline Debate/Cost.lean]: Can you use the upstream `Nat.ceil_lt_add_one` instead of this lemma?

File: Debate/Details.lean [hunk 2]
```diff
@@ -645,3 +653,32 @@ theorem soundness_p (o : Oracle) (L : o.lipschitz t k) (eve : Alice)
   · apply add_le_add_right m
   · linarith [p.bw]
   · apply pow_nonneg; linarith [p.q1]
+
+/-- Default valid parameters (not tuned) -/
+def defaults (k : ℝ) (t : ℕ) (k0 : 0 < k) : Params (2/3) (3/5) k t where
```
Comment (eric-wieser) [inline Debate/Details.lean]: Perhaps `Params.defaults` for clarity?

File: Debate/Cost.lean [hunk 3]
```diff
@@ -109,6 +111,44 @@ theorem bob_debate_cost (o : Oracle) (alice : Alice) (vera : Vera) (t : ℕ):
...
+  norm_num
+  simp only [mul_comm (Real.log _)]
+  refine le_trans (Nat.ceil_le_add_one ?_) (le_refl _)
+  exact mul_nonneg (by positivity) (Real.log_nonneg (by linarith))
+
+/-- Bob makes `O(k^2 t log t)` queries with default parameters -/
+theorem bob_fast (k : ℝ) (k0 : 0 < k) (t : ℕ) (alice : Alice) (vera : Vera) :
+    let p := defaults k t k0
+    (debate alice (bob p.s p.b p.q) vera t).cost' o BobId ≤
+      (t+1) * (20000 / 9 * k^2 * Real.log (200 * (t+1)) + 1) := by
```
Comment (eric-wieser) [inline Debate/Cost.lean:134]: Is it feasible to write a version of this using asymptotics notation, that more closely resembles the docstring? (happy to merge without this)

File: Debate/Cost.lean [hunk 4]
```diff
@@ -175,3 +215,25 @@ theorem vera_debate_cost (o : Oracle) (alice : Alice) (bob : Bob) (t : ℕ):
     · simp only [Comp.cost_bind, Comp.cost_allow_all, vera_cost, Comp.prob_allow_all,
         Comp.cost_pure, exp_const, add_zero, le_refl]
     · simp only [Comp.cost_pure, Nat.cast_nonneg]
+
+/-- A calculation used in `vera_fast` -/
+lemma log_mul_le : Real.log 200 * 20000 ≤ 106000 := by
```
Comment (eric-wieser) [inline Debate/Cost.lean]:
```suggestion
lemma log_200_mul_20000_le : Real.log 200 * 20000 ≤ 106000 := by
```

**Review by eric-wieser — APPROVED** [review]
Some minor comments

File: Debate/Cost.lean (hunk 1 above)
Comment (girving) [reply Debate/Cost.lean]: Done.

File: Debate/Cost.lean (hunk 3 above)
Comment (girving) [reply Debate/Cost.lean:134]: I want the explicit constant version: https://x.com/geoffreyirving/status/1805903946629701764

Comment (girving) [reply Debate/Cost.lean:134]: I want the explicit constant version: https://x.com/geoffreyirving/status/1805903946629701764

File: Debate/Cost.lean (hunk 4 above)
Comment (girving) [reply Debate/Cost.lean]: Done.

File: Debate/Details.lean (hunk 2 above)
Comment (girving) [reply Debate/Details.lean]: Done.

File: Debate/Cost.lean (hunk 3 above)
Comment (girving) [reply Debate/Cost.lean:134]: Ah, you probably mean *additionally* have the asymptotic versions. I’d like to triage that away for now if possible.

File: Debate/Cost.lean [hunk 5]
```diff
@@ -175,3 +210,25 @@ theorem vera_debate_cost (o : Oracle) (alice : Alice) (bob : Bob) (t : ℕ):
     · simp only [Comp.cost_bind, Comp.cost_allow_all, vera_cost, Comp.prob_allow_all,
         Comp.cost_pure, exp_const, add_zero, le_refl]
     · simp only [Comp.cost_pure, Nat.cast_nonneg]
+
+/-- A calculation used in `vera_fast` -/
+lemma  log_200_mul_20000_le : Real.log 200 * 20000 ≤ 106000 := by
```
Comment (eric-wieser) [inline Debate/Cost.lean]:
```suggestion
lemma log_200_mul_20000_le : Real.log 200 * 20000 ≤ 106000 := by
```

**Pull request merged**